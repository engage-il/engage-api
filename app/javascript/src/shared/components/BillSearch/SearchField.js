// @flow
import * as React from 'react'
import styled from 'react-emotion'
import { Link } from '../Link'
import { colors, mixins } from 'shared/styles'

type Props = {
  value: string,
  onChange: (string) => void
}

type State = {
  isFocused: boolean
}

export class SearchField extends React.Component<*, Props, State> {
  state = {
    isFocused: false
  }

  // events
  inputDidChange = (event: { target: { value: string } }) => {
    const { onChange } = this.props
    onChange(event.target.value)
  }

  inputDidChangeFocus = (isFocused: boolean) => () => {
    this.setState({ isFocused })
  }

  // lifecycle
  render () {
    const { value } = this.props
    const { isFocused } = this.state

    return (
      <Container>
        <Field isFocused={isFocused}>
          <Input
            type='text'
            name='search-field'
            value={value}
            placeholder={`health care, HB2364`}
            onChange={this.inputDidChange}
            onFocus={this.inputDidChangeFocus(true)}
            onBlur={this.inputDidChangeFocus(false)}
          />
          <Button
            onClick={() => {}}
            children='Search'
          />
        </Field>
      </Container>
    )
  }
}

const Container = styled.div`
  ${mixins.flexRow};

  justify-content: center;
  align-items: center;
  height: 190px;
  background-color: ${colors.primary};
`

const Field = styled.div`
  ${mixins.flexRow};
  ${mixins.pageWidth};
  ${mixins.pageMargin};

  flex: 1;
  height: 70px;
  border: solid ${colors.white};
  border-width: ${({ isFocused }: State) => isFocused ? 2 : 1}px;
  border-radius: 10px;
  overflow: hidden;
  transform: perspective(200px) translateZ(${({ isFocused }: State) => isFocused ? 1 : 0}px);
  font-size: 24px;
  color: ${colors.white};
  transition: border-width 0.15s, transform 0.15s;
`

const Input = styled.input`
  flex: 1;
  padding: 0 30px;
  color: inherit;
  background-color: transparent;
  font-size: inherit;

  ::placeholder {
    color: ${colors.white};
    opacity: 0.5;
  }
`

const Button = styled(Link)`
  ${mixins.flexRow};
  ${mixins.flexCenter};
  ${mixins.fonts.bold};

  width: 215px;
  color: ${colors.primary};
  background-color: ${colors.white};
  text-transform: uppercase;
`
