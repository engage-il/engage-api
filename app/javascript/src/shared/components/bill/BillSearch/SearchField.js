// @flow
import * as React from 'react'
import styled, { css } from 'react-emotion'
import { Link } from '@/components'
import { colors, mixins } from '@/styles'
import type { SearchParams } from '@/types'

type Props = {
  params: SearchParams,
  onChange: (SearchParams) => void
}

type State = {
  isFocused: boolean
}

export class SearchField extends React.Component<Props, State> {
  state = {
    isFocused: false
  }

  // events
  didChangeQuery = ({ target }: SyntheticInputEvent<*>) => {
    const { params, onChange } = this.props

    onChange({
      ...params,
      query: target.value
    })
  }

  didChangeFocus = (isFocused: boolean) => () => {
    this.setState({ isFocused })
  }

  // lifecycle
  render () {
    const { params } = this.props
    const { isFocused } = this.state

    return (
      <Search>
        <FieldContainer>
          <Field isFocused={isFocused}>
            <Input
              type='text'
              name='search-field'
              value={params.query}
              placeholder={`health care, HB2364`}
              onChange={this.didChangeQuery}
              onFocus={this.didChangeFocus(true)}
              onBlur={this.didChangeFocus(false)}
            />
            <Button
              onClick={() => {}}
              children='Search'
            />
          </Field>
        </FieldContainer>
      </Search>
    )
  }
}

const Search = styled.div`
  ${mixins.flexRow};

  justify-content: center;
  align-items: center;
  height: 150px;
  background-color: ${colors.primary};

  ${mixins.mobile`
    height: 120px;
  `}
`

const FieldContainer = styled.div`
  ${mixins.flexRow};
  ${mixins.pageWidth};

  flex: 1;
  height: 56px;
  padding: 0 70px;

  ${mixins.mobile`
    padding: 0 30px;
  `}
`

const focused = css`
  border-width: 2px;
  transform: perspective(200px) translateZ(1px);
`

const Field = styled.div`
  ${mixins.flexRow};
  ${mixins.font.sizes.h5};

  flex: 1;
  border: solid ${colors.white};
  border-width: 1px;
  border-radius: 10px;
  overflow: hidden;
  transform: perspective(200px) translateZ(0px);
  color: ${colors.white};
  transition: border-width 0.15s, transform 0.15s;

  ${({ isFocused }) => isFocused && focused}
`

const Input = styled.input`
  flex: 1;
  padding: 0 25px;

  &:focus {
    cursor: text;
  }

  ::placeholder {
    color: ${colors.white};
    opacity: 0.5;
  }

  ${mixins.mobile`
    padding: 0 15px;
  `}
`

const Button = styled(Link)`
  ${mixins.flexRow};
  ${mixins.flexCenter};
  ${mixins.font.bold};

  width: 215px;
  color: ${colors.primary};
  background-color: ${colors.white};
  text-transform: uppercase;

  ${mixins.mobile`
    width: 90px;
  `}
`
