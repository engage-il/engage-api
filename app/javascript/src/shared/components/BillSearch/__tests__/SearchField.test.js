/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import { defaultsDeep } from 'lodash'
import { SearchField } from '../SearchField'

// subject
let subject

const defaults = {
  value: 'test-value',
  onChange: jest.fn()
}

function loadSubject (props = {}) {
  subject = shallow(<SearchField {...defaultsDeep(props, defaults)} />)
}

// specs
afterEach(() => {
  subject = null
})

describe('#render', () => {
  it('shows the search field', () => {
    loadSubject()
    expect(subject).toMatchSnapshot()
  })
})

describe('entering a value', () => {
  it('notifies its parent', () => {
    loadSubject()
    subject.instance().inputDidChange({ target: { value: 'foo' } })
    expect(defaults.onChange).toHaveBeenCalledWith('foo')
  })
})

describe('focusing the input', () => {
  it('focuses the component', () => {
    loadSubject()
    subject.instance().inputDidChangeFocus(true)()
    expect(subject).toHaveState('isFocused', true)
  })
})
