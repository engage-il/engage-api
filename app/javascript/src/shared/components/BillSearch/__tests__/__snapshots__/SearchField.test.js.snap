// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render shows the search field 1`] = `
<Styled(div)>
  <Styled(div)>
    <Styled(div)
      isFocused={false}
    >
      <Styled(input)
        name="search-field"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="health care, HB2364"
        type="text"
        value="test-value"
      />
      <Styled(Link)
        onClick={[Function]}
      >
        Search
      </Styled(Link)>
    </Styled(div)>
  </Styled(div)>
</Styled(div)>
`;
